### PR TITLE
DPIB-496: AI Feature integrations.

### PR DESCRIPTION
**Pull request by bohdanyevtukhov**
- Developer(s): @bohdanyevtukhov, @mantasmuliar 

- Ticket: https://spryker.atlassian.net/browse/DPIB-496

- Release Group: https://release.spryker.com/release-groups/view/5536

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   ProductManagementAi        | minor(new)                 |    OpenAi                   |

-----------------------------------------

##### Change log

Initial Release

- Introduced ProductManagementAi module to provide product management with AI capabilities.

### CODE REVIEW

File: assets/Zed/js/ai-category-suggestion.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import { AiProductManagement } from './ai-product-management';
```
Comment (dmiseev) [inline assets/Zed/js/ai-category-suggestion.js:1]: Ask @tolerants for FE review.

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
+{
+    "name": "spryker-eco/product-management-ai",
```
Comment (dmiseev) [inline composer.json:2]: self-check: check dependencies before R4R label

File: src/SprykerEco/Shared/ProductManagementAi/Transfer/product_management_ai.transfer.xml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0"?>
```
Comment (dmiseev) [inline src/SprykerEco/Shared/ProductManagementAi/Transfer/product_management_ai.transfer.xml:1]: vendor/bin/spryker-dev-console dev:validate-module-transfers -m SprykerEco.ProductManagementAi -f

File: src/SprykerEco/Zed/ProductManagementAi/Business/Builder/PromptBuilder.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\ProductManagementAi\Business\Builder;
+
+use Locale;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/Builder/PromptBuilder.php]: get rid?

File: src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryProposer.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\ProductManagementAi\Business\Category;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryProposer.php]: We have to be consistent with business model folder names: lets group it by action (Expander, Proposer, Creator, etc).

File: src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryProposer.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+use Generated\Shared\Transfer\OpenAiChatRequestTransfer;
+use SprykerEco\Zed\ProductManagementAi\Dependency\Client\ProductManagementAiToOpenAiClientInterface;
+use SprykerEco\Zed\ProductManagementAi\Dependency\Service\ProductManagementAiToUtilEncodingServiceInterface;
+
+class CategoryProposer implements CategoryProposerInterface
+{
+    /**
+     * @var string
+     */
+    protected const PRODUCT_CATEGORY_SUGGESTION_PROMPT = 'Based on the provided product name and description, suggest the most fitting product categories from the existing categories list for optimal placement in an e-commerce store.
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryProposer.php]: should be configurable, please move to config.

File: src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryReader.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\ProductManagementAi\Business\Category;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryReader.php]: Same, move to Reader subfolder.

File: src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryReader.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        $categoryCollectionTransfer = $this->categoryFacade
+            ->getAllCategoryCollection($this->localeFacade->getCurrentLocale());
+
+        $categories = [];
+        foreach ($categoryCollectionTransfer->getCategories() as $categoryTransfer) {
+            $categories[$categoryTransfer->getName()] = $categoryTransfer->getIdCategoryOrFail();
+        }
+
+        return $categories;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/Category/CategoryReader.php]: Method SprykerEco\Zed\ProductManagementAi\Business\Category\CategoryReader::getCategories() should return array<string, int> but returns array<int|string, mixed>.

File: src/SprykerEco/Zed/ProductManagementAi/Business/ProductManagementAiFacadeInterface.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+use Generated\Shared\Transfer\AiTranslatorRequestTransfer;
+use Generated\Shared\Transfer\AiTranslatorResponseTransfer;
+use Generated\Shared\Transfer\OpenAiChatResponseTransfer;
+
+interface ProductManagementAiFacadeInterface
+{
+    /**
+     * Specification:
+     * - propose category suggestions based on product name, description and existing categories
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/ProductManagementAiFacadeInterface.php]:
```suggestion
     * - Proposes category suggestions based on product name, description and existing categories.
```

File: src/SprykerEco/Zed/ProductManagementAi/Business/Translator/StorageKeyBuilder/StorageKeyBuilder.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEco\Zed\ProductManagementAi\Business\Translator\StorageKeyBuilder;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/Translator/StorageKeyBuilder/StorageKeyBuilder.php]: Please avoid nested folders in Business layer.

File: src/SprykerEco/Zed/ProductManagementAi/Business/Translator/TranslatorCacheAware.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * @param \Generated\Shared\Transfer\AiTranslatorRequestTransfer $aiTranslatorRequestTransfer
+     *
+     * @return \Generated\Shared\Transfer\AiTranslatorResponseTransfer
+     */
+    public function translate(AiTranslatorRequestTransfer $aiTranslatorRequestTransfer): AiTranslatorResponseTransfer
+    {
+        $cacheKey = $this->storageKeyBuilder->buildStorageKey($aiTranslatorRequestTransfer);
+        if (!$aiTranslatorRequestTransfer->getInvalidateCache()) {
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Business/Translator/TranslatorCacheAware.php]: Do we have a requirement for caching?

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Expander/ImageAltTextProductFormExpander.php [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+{
+    /**
+     * @var string
+     */
+    protected const SUBFORM_IMAGE_SET_PREFIX = 'image_set_';
+
+    /**
+     * @var string
+     */
+    public const FIELD_ALT_TEXT = 'alt_text';
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Expander/ImageAltTextProductFormExpander.php]:
```suggestion
    protected const FIELD_ALT_TEXT = 'alt_text';
```

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Expander/ImageAltTextProductFormExpander.php [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        $builder->addEventListener(
+            FormEvents::POST_SET_DATA,
+            function (FormEvent $event): void {
+                $form = $event->getForm();
+
+                $localeCollection = $this->localeFacade->getLocaleCollection();
+                $localeCollection[] = (new LocaleTransfer())->setLocaleName('default');
+
+                foreach ($localeCollection as $localeTransfer) {
+                    $imageSetCollection = $form->get($this->getLocalizedImageSetFormName($localeTransfer));
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Expander/ImageAltTextProductFormExpander.php]: form could be changed, please ensure that form exists (use `has()` method)

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractAfterUpdatePlugin.php [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        $assignedCategoryIdsToDelete = array_diff($assignedCategoryIds, $categoryIdsToAssign);
+        foreach ($assignedCategoryIdsToSave as $idCategory) {
+            $productCategoryFacade->createProductCategoryMappings($idCategory, [$productAbstractTransfer->getIdProductAbstract()]);
+        }
+
+        foreach ($assignedCategoryIdsToDelete as $idCategory) {
+            $productCategoryFacade->removeProductCategoryMappings($idCategory, [$productAbstractTransfer->getIdProductAbstract()]);
+        }
+
+        return $productAbstractTransfer;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractAfterUpdatePlugin.php]: please move business logic to business model (in Communication layer)

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractAfterUpdatePlugin.php [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * @param int $idProductAbstract
+     *
+     * @return array
+     */
+    protected function getProductCategoryIds(int $idProductAbstract): array
+    {
+        // Use Facade
+        return SpyProductCategoryQuery::create()
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractAfterUpdatePlugin.php]: Use Facade method instead Query

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractPostCreatePlugin.php [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * @method \SprykerEco\Zed\ProductManagementAi\Communication\ProductManagementAiCommunicationFactory getFactory()
+ * @method \SprykerEco\Zed\ProductManagementAi\ProductManagementAiConfig getConfig()
+ * @method \SprykerEco\Zed\ProductManagementAi\Business\ProductManagementAiFacadeInterface getFacade()
+ */
+class ProductCategoryProductAbstractPostCreatePlugin extends AbstractPlugin implements ProductAbstractPostCreatePluginInterface
+{
+    /**
+     * {@inheritDoc}
+     * Specification:
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractPostCreatePlugin.php]: Get rid

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractPostCreatePlugin.php [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+     * @return \Generated\Shared\Transfer\ProductAbstractTransfer
+     */
+    public function postCreate(ProductAbstractTransfer $productAbstractTransfer): ProductAbstractTransfer
+    {
+        $productCategoryFacade = new ProductCategoryFacade();
+        foreach ($productAbstractTransfer->getCategoryIds() as $idCategory) {
+            $productCategoryFacade->createProductCategoryMappings($idCategory, [$productAbstractTransfer->getIdProductAbstract()]);
+        }
+
+        return $productAbstractTransfer;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/Product/ProductCategoryProductAbstractPostCreatePlugin.php]: To business model

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ImageAltTextProductAbstractFormExpanderPlugin.php [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * @method \SprykerEco\Zed\ProductManagementAi\ProductManagementAiConfig getConfig()
+ * @method \SprykerEco\Zed\ProductManagementAi\Business\ProductManagementAiFacadeInterface getFacade()
+ * @method \SprykerEco\Zed\ProductManagementAi\Communication\ProductManagementAiCommunicationFactory getFactory()
+ */
+class ImageAltTextProductAbstractFormExpanderPlugin extends AbstractPlugin implements ProductAbstractFormExpanderPluginInterface, ProductConcreteFormExpanderPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     *
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ImageAltTextProductAbstractFormExpanderPlugin.php:26]: Spec is missing.

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ImageAltTextProductConcreteEditFormExpanderPlugin.php [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * @method \SprykerEco\Zed\ProductManagementAi\ProductManagementAiConfig getConfig()
+ * @method \SprykerEco\Zed\ProductManagementAi\Business\ProductManagementAiFacadeInterface getFacade()
+ * @method \SprykerEco\Zed\ProductManagementAi\Communication\ProductManagementAiCommunicationFactory getFactory()
+ */
+class ImageAltTextProductConcreteEditFormExpanderPlugin extends AbstractPlugin implements ProductConcreteEditFormExpanderPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     *
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ImageAltTextProductConcreteEditFormExpanderPlugin.php:24]: Same

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ImageAltTextProductConcreteFormExpanderPlugin.php [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * @method \SprykerEco\Zed\ProductManagementAi\ProductManagementAiConfig getConfig()
+ * @method \SprykerEco\Zed\ProductManagementAi\Business\ProductManagementAiFacadeInterface getFacade()
+ * @method \SprykerEco\Zed\ProductManagementAi\Communication\ProductManagementAiCommunicationFactory getFactory()
+ */
+class ImageAltTextProductConcreteFormExpanderPlugin extends AbstractPlugin implements ProductConcreteFormExpanderPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     *
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ImageAltTextProductConcreteFormExpanderPlugin.php:24]: Same

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ProductCategoryAbstractFormExpanderPlugin.php [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+                if (!$data || !isset($data[self::KEY_ID_PRODUCT_ABSTRACT])) {
+                    return;
+                }
+
+                $data[ProductAbstractTransfer::CATEGORY_IDS] = $this->getProductCategoryIds($data[self::KEY_ID_PRODUCT_ABSTRACT]);
+                $event->setData($data);
+            },
+        );
+
+        return $builder;
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ProductCategoryAbstractFormExpanderPlugin.php]: It should be ...FormExpander, avoid business logic in Plugins.

File: src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ProductCategoryAbstractFormExpanderPlugin.php [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * @param int $idProductAbstract
+     *
+     * @return array<int>
+     */
+    protected function getProductCategoryIds(int $idProductAbstract): array
+    {
+        $currentLocaleId = $this->getFactory()->getLocaleFacade()->getCurrentLocale()->getIdLocale();
+        // Use Facade
+        $categoryCollection = SpyCategoryQuery::create()
```
Comment (dmiseev) [inline src/SprykerEco/Zed/ProductManagementAi/Communication/Plugin/ProductManagement/ProductCategoryAbstractFormExpanderPlugin.php]: Use Facade

File: assets/Zed/js/ai-category-suggestion.js (hunk 1 above)
Comment (supproduction) [reply assets/Zed/js/ai-category-suggestion.js:1]: we can push to the eco modules without FE approval

File: composer.json [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+    "type": "library",
+    "description": "ProductManagementAi module",
+    "license": "MIT",
+    "require": {
+        "php": ">=8.1",
+        "spryker-eco/open-ai": "dev-feature/demo/dev-ai-integrations",
```
Comment (dmiseev) [inline composer.json]: Before release: `^1.0.0`

File: composer.json [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+        "spryker/propel-orm": "Used for Propel schema generation."
+    },
+    "autoload": {
+        "psr-4": {
+            "SprykerEco\\": "src/SprykerEco/"
+        }
+    },
+    "minimum-stability": "dev",
+    "prefer-stable": true,
+    "repositories": [
```
Comment (dmiseev) [inline composer.json]: Before release: get rid

**Review by dmiseev — APPROVED** [review]

**Review by dmiseev — APPROVED** [review]

**Pull request merged**